examples/ecs/custom_query_param.rs

```diff
@@ -0,0 +1,189 @@
+        )
+        .add_system(
+            print_components_read_only
+                .label("print_components_read_only")
+                .after("print_components_iter"),
+        )
+        .add_system(print_components_tuple.after("print_components_read_only"))
+        .run();
+}
+
+#[derive(Component, Debug)]
+struct ComponentA;
+#[derive(Component, Debug)]
+struct ComponentB;
+#[derive(Component, Debug)]
+struct ComponentC;
+#[derive(Component, Debug)]
+struct ComponentD;
+#[derive(Component, Debug)]
+struct ComponentZ;
+
+#[derive(Fetch)]
+struct CustomQuery<'w, T: Component + Debug, P: Component + Debug> {
+    entity: Entity,
+    // `Mut<'w, T>` is a necessary replacement for `&'w mut T`
+    a: Mut<'w, ComponentA>,
+    b: Option<Mut<'w, ComponentB>>,
+    nested: NestedQuery<'w>,
+    generic: GenericQuery<'w, T, P>,
+    #[allow(dead_code)]
+    empty: EmptyQuery<'w>,
+}
+
+// This is a valid query as well, which would iterate over every entity.
+#[derive(Fetch)]
+struct EmptyQuery<'w> {
+    _w: std::marker::PhantomData<&'w ()>,
+}
+
+#[derive(Fetch, Debug)]
+#[read_only_derive(Debug)]
+#[allow(dead_code)]
+struct NestedQuery<'w> {
+    c: &'w ComponentC,
+    d: Option<&'w ComponentD>,
+}
+
+#[derive(Fetch, Debug)]
+#[read_only_derive(Debug)]
+#[allow(dead_code)]
+struct GenericQuery<'w, T: Component, P: Component> {
+    generic: (&'w T, &'w P),
+}
+
+#[derive(FilterFetch)]
+struct QueryFilter<T: Component, P: Component> {
+    _c: With<ComponentC>,
+    _d: With<ComponentD>,
+    _or: Or<(Added<ComponentC>, Changed<ComponentD>, Without<ComponentZ>)>,
+    _generic_tuple: (With<T>, With<P>),
+    _tp: PhantomData<(T, P)>,
+}
+
+fn spawn(mut commands: Commands) {
+    commands
+        .spawn()
+        .insert(ComponentA)
+        .insert(ComponentB)
+        .insert(ComponentC)
+        .insert(ComponentD);
+}
+
+fn print_components_iter_mut(
+    mut query: Query<CustomQuery<ComponentC, ComponentD>, QueryFilter<ComponentC, ComponentD>>,
+) {
+    println!("Print components (iter_mut):");
+    for e in query.iter_mut() {
+        println!("Entity: {:?}", e.entity);
+        println!("A: {:?}", e.a);
+        println!("B: {:?}", e.b);
+        println!("Nested: {:?}", e.nested);
+        println!("Generic: {:?}", e.generic);
+    }
+    println!();
+}
+
+fn print_components_iter(
+    query: Query<CustomQuery<ComponentC, ComponentD>, QueryFilter<ComponentC, ComponentD>>,
+) {
+    println!("Print components (iter):");
+    for e in query.iter() {
+        // Note that the actual type is different when you iterate with `iter`.
+        let e: CustomQueryReadOnly<'_, _, _> = e;
+        println!("Entity: {:?}", e.entity);
+        println!("A: {:?}", e.a);
+        println!("B: {:?}", e.b);
+        println!("Nested: {:?}", e.nested);
+        println!("Generic: {:?}", e.generic);
+    }
+    println!();
+}
+
+// If you are going to use a query only for reading, you can mark it with `read_only` attribute
```

Comment:
IMO we also need to explain to the end user *why* they might ever want to do this here. IIRC it'll enable `ReadOnlySystemParamFetch`, which will then accelerate / simplify some of the conflict checking?